### ISSUE TITLE
Potential object leakage in 0.5.x

### ISSUE THREAD

**Issue by jsvd** [cheald/manticore#45]
while investigation the root cause of https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/392, I'm finding Manticore to have a suspicious behavior in terms of allocation and obj evictions.

I used the following script and a webserver open in 9200:

```ruby
require 'manticore'

m = Manticore::Client.new :hosts => ["http://localhost:9200/"]

def gen_stuff
  10000.times { Manticore::Client.new :hosts => ["http://localhost:9200/"] }
  true
end

gen_stuff

puts "done, sleeping"

sleep
```
I ran this script with jruby 1.7.23 and manticore 0.5.2, then I used Yourkit to analyse the memory. First I forced a GC then updated the count of objects, which should only show one instance of the HttpClient. however:

<img width="951" alt="screen shot 2016-03-14 at 19 17 48" src="https://cloud.githubusercontent.com/assets/31809/13756796/8d7c2b8a-ea19-11e5-81c5-2ee52b44aba2.png">

Trying with manticore 0.4.4 shows only 2 instances:

<img width="944" alt="screen shot 2016-03-14 at 19 18 12" src="https://cloud.githubusercontent.com/assets/31809/13756817/9ede722a-ea19-11e5-990a-d609c8ac31b5.png">

There are many other instances related to HttpClient that also survive, so I'm wondering if there's some issue in freeing up objects after leaving scope?

**Comment by cheald** [comment]
This is probably the ConnectionManager and/or ThreadPool that Manticore creates. They are torn down in `at_exit` blocks, but aren't managed via object finalizers. That probably needs to be fixed, but it should be noted that Manticore clients are relatively expensive to create (because a new Client has to allocate all the HttpClient stuff under the hood) and should be created once and then reused if at all possible.

I'll see if I can ID which instances are hanging around, but if you look for `at_exit` in client.rb, it's probably the things managed by those blocks. :)

**Comment by cheald** [comment]
If you could, try testing by calling `client.close`, as well; in theory that should clean up the underlying Java object allocations.

The allocation stuff is a little tricky here, because pools are intended to be created and reused across Manticore client instances, so we can't assume that a client being collected should result in the closure of its associated pool, and AFAIK we don't get reference counting across the Java<->Ruby boundary, so it may be that clients require explicit destruction to clean up properly. At a minimum though, I need to test and document this.

**Comment by cheald** [comment]
Irony of ironies, it's the `at_exit` hooks holding onto those HttpClient references even after the Client has been collected. Those were added because of things like #26. I'll figure out how to support both cases.

**Comment by cheald** [comment]
Okay, I think b774434f90dd3a994475236bdb5596c67e12575f should address both cases. I've removed the `at_exit` hooks and replaced them with finalizers, which should result in the proper behavior. I've tested this with the following script:

```
require_relative 'lib/manticore'

100.times do
  client = Manticore::Client.new :hosts => ["http://localhost:9200/"]
  GC.start
end

client = Manticore::Client.new :hosts => ["http://localhost:9200/"]
puts "okay, waiting"
gets
```

which I then measure with:

    jmap -histo:live `pgrep -f object.rb` | grep HttpClient

With this fix in place, the internal client objects are cleaned up after each run. I also put together a test servlet

config.ru

```
require 'rack'
require 'bundler/setup'
require 'manticore'

app = Proc.new do |env|
  client = Manticore::Client.new
  str = client.get("http://mashable.com").body
  ['200', {'Content-Type' => 'text/html'}, [str]]
end

run app
```

with Gemfile:

```
source 'https://rubygems.org'
gem 'manticore', git: "../.git", branch: "finalizers"
gem 'warbler'
```

packed it into a WAR, deployed it on Wildfly, and observed HttpClient internals via jmap again. HttpClient internals were properly cleaned up both between requests and on redeployment of the WAR.

@jsvd If you'd like to test that fix and ensure that it's doing what you'd expect, I'll go ahead and close this out and ship a new version of the gem.

**Comment by jsvd** [comment]
oh wow, so the at_exit blocks were creating a closure that kept a reference to the objects...

nice catch @cheald, I ran my little test with the patch and it works perfectly :+1: 

thanks for the amazingly fast response and fix!

**Comment by cheald** [comment]
Released 0.5.4
